### PR TITLE
Independent spinbox arrow step precision

### PR DESCRIPTION
**Pull request by EIREXE**
Rebase of #33743

_Bugsquad Edit:_ Implements godotengine/godot-proposals#226.

### CODE REVIEW

**Review by KoBeWi — APPROVED** [review]
Instead of repeating `MAX(get_step(), get_arrow_step())` in `gui_input()`, you could make a variable at the beginning: `double step = MAX(get_step(), get_arrow_step())`. Other than that the PR is fine.

File: doc/classes/SpinBox.xml [hunk 1]
```diff
@@ -47,6 +47,9 @@
 		<member name="align" type="int" setter="set_align" getter="get_align" enum="LineEdit.Align" default="0">
 			Sets the text alignment of the [SpinBox].
 		</member>
+		<member name="arrow_step" type="float" setter="set_arrow_step" getter="get_arrow_step" default="0.0">
+			If greater than 0, [code]value[/code] will always be rounded to a multiple of [code]arrow_step[/code] when interacting with the arrow buttons of the [SpinBox].
```
Comment (KoBeWi) [inline doc/classes/SpinBox.xml:51]:
```suggestion
			If greater than [code]0[/code], [code]value[/code] will always be rounded to a multiple of [code]arrow_step[/code] when interacting with the arrow buttons of the [SpinBox].
```

**Comment by EIREXE** [comment]
@KoBeWi should be ok now

File: scene/gui/spin_box.cpp [hunk 2]
```diff
@@ -167,7 +169,7 @@ void SpinBox::gui_input(const Ref<InputEvent> &p_event) {
 		if (drag.enabled) {
 			drag.diff_y += mm->get_relative().y;
 			float diff_y = -0.01 * Math::pow(ABS(drag.diff_y), 1.8f) * SIGN(drag.diff_y);
-			set_value(CLAMP(drag.base_val + get_step() * diff_y, get_min(), get_max()));
+			set_value(CLAMP(drag.base_val + step * ceil(diff_y), get_min(), get_max()));
```
Comment (KoBeWi) [inline scene/gui/spin_box.cpp:172]: What is the purpose of newly added `ceil()`?

Comment (EIREXE) [reply scene/gui/spin_box.cpp:172]: > What is the purpose of newly added `ceil()`?

I have no idea, it was in the original PR, I can remove it if needed

Comment (KoBeWi) [reply scene/gui/spin_box.cpp:172]: Yeah, it doesn't seen necessary. It can be removed.

**Comment by reduz** [comment]
I understand why this is needed, but I would probably do the following changes:
* Rename "arrow_step" to "custom_arrow_step"
* Would use it when != 0.0 , not with max/min

I think with these two changes, the workings should be more intuitive.

**Comment by EIREXE** [comment]
I think it should be ok now @reduz

File: scene/gui/spin_box.cpp [hunk 3]
```diff
@@ -294,6 +297,14 @@ void SpinBox::apply() {
 	_text_submitted(line_edit->get_text());
 }
 
+void SpinBox::set_custom_arrow_step(double p_custom_arrow_step) {
+	custom_arrow_step = p_custom_arrow_step;
+}
+
+double SpinBox::get_custom_arrow_step() const {
+	return custom_arrow_step;
+}
```
Comment (akien-mga) [inline scene/gui/spin_box.cpp:306]: These should also be renamed to remove the `custom_` prefix I think (including the member var). Remember to update the binding and the docs too.

Comment (EIREXE) [reply scene/gui/spin_box.cpp:306]: What do you mean remove? Isn't the point of the changes requested by juan that we ADD the `custom_` prefix?

Comment (akien-mga) [reply scene/gui/spin_box.cpp:306]: Ah yes. Then you just forgot to update the property name :)

Comment (EIREXE) [reply scene/gui/spin_box.cpp:306]: Ah I see, yeah that sounds likely, I will fix it now.

**Review by akien-mga — APPROVED** [review]
Looks good to me.

**Comment by akien-mga** [comment]
Thanks!

**Comment by esijg** [comment]
Thank you for this! Love seeing someone taking good care of this feature :)

**Pull request merged**